src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,290 @@
 
 module Data.Nat.Primality where
 
-open import Data.Empty using (⊥)
+open import Data.Bool.Base using (Bool; true; false; not; T)
 open import Data.Nat.Base
 open import Data.Nat.Divisibility
 open import Data.Nat.GCD using (module GCD; module Bézout)
 open import Data.Nat.Properties
-open import Data.Product.Base using (∃; _×_; map₂; _,_)
-open import Data.Sum.Base using (_⊎_; inj₁; inj₂)
+open import Data.Product.Base using (_×_; map₂; _,_; proj₂)
+open import Data.Sum.Base using (_⊎_; inj₁; inj₂; [_,_]′)
 open import Function.Base using (flip; _∘_; _∘′_)
 open import Relation.Nullary.Decidable as Dec
-  using (yes; no; from-yes; ¬?; decidable-stable; _×-dec_; _→-dec_)
+  using (yes; no; from-yes; from-no; ¬?; decidable-stable; _×-dec_; _⊎-dec_; _→-dec_)
 open import Relation.Nullary.Negation using (¬_; contradiction)
-open import Relation.Unary using (Decidable)
+open import Relation.Unary using (Pred; Decidable; IUniversal; Satisfiable)
 open import Relation.Binary.PropositionalEquality
-  using (refl; cong)
+  using (_≡_; _≢_; refl; cong)
 
 private
   variable
-    n : ℕ
+    d k m n p : ℕ
+
+------------------------------------------------------------------------
+-- Definitions for `Data.Nat.Base`
+
+pattern 2+ n = suc (suc n)
+
+trivial : ℕ → Bool
+trivial 0      = true
+trivial 1      = true
+trivial (2+ _) = false
+
+Unit NonUnit NonTrivial : Pred ℕ _
+Unit       = T ∘ (_≡ᵇ 1)
+NonUnit    = T ∘ not ∘ (_≡ᵇ 1)
+NonTrivial = T ∘ not ∘ trivial
+
+instance
+  nonUnit[0] : NonUnit 0
+  nonUnit[0] = _
+
+  nonTrivial⇒nonUnit : .{{NonTrivial n}} → NonUnit n
+  nonTrivial⇒nonUnit {n = 2+ _} = _
+
+nonUnit⇒≢1 : .{{NonUnit n}} → n ≢ 1
+nonUnit⇒≢1 ⦃()⦄ refl
+instance
+  nonTrivial : NonTrivial (2+ n)
+  nonTrivial = _
+
+nonTrivial⇒≢1 : .{{NonTrivial n}} → n ≢ 1
+nonTrivial⇒≢1 ⦃()⦄ refl
+
+nonTrivial⇒nonZero : .{{NonTrivial n}} → NonZero n
+nonTrivial⇒nonZero {n = 2+ k} = _
```

Comment:
A thing I don't understand is that my instinct woud have been to make this an `instance` declaration... but that appears to cause some subsequent proofs to *fail* (I think; but running these things interactively doesn't always give accurate feedback, again seemingly because of hidden cached state in the editor/proof engine; but maybe that's faulty rationalisation on my part?)